pkg/util/audit/audit.go

```diff
@@ -0,0 +1,158 @@
+	// in an implementation-independent way.
+	ifxAuditFlags = 257
+
+	// ifxAuditPopSample is the effective percentage sample rate for this event at the
+	// time it was generated by a client (either the default value specified in
+	// the event definition, system default or a dynamic override from a cloud
+	// settings service).
+	// not used by Azure
+	ifxAuditPopSample = 0.0
+)
+
+var (
+	// epoch is an unique identifier associated with the current session of the
+	// telemetry library running on the platform. It must be stable during a
+	// session, and has no implied ordering across sessions.
+	epoch string
+
+	// sequence field used to track absolute order of uploaded events, per session.
+	// it is reset when the ARO component is restarted.
+	seqNum      uint64
+	seqNumMutex *sync.Mutex
+)
+
+func init() {
+	epoch = uuid.NewV4().String()
+	seqNum = 1
+	seqNumMutex = new(sync.Mutex)
+}
+
+// EmitRPLog is used by aro-rp to emit audit logs.
+// Caller is responsible for providing the logrus.Entry object that contains
+// the appropriate formatter, hooks, log level etc.
+// All audit logs will be emitted at the INFO level.
+func EmitRPLog(entry logrus.Entry, log *Log) error {
+	return emitLog(entry, log, auditSourceRP)
+}
+
+func emitLog(entry logrus.Entry, log *Log, source string) error {
+	payload := log.toPayload()
+
+	fields := logrus.Fields{}
+	fields[auditMetadataCategory] = payload.Category
+	fields[auditMetadataCreatedTime] = time.Now().UTC().Format(time.RFC3339)
+	fields[auditMetadataLogKind] = ifxAuditLogKind
+	fields[auditMetadataOperation] = payload.OperationName
+	fields[auditMetadataResult] = payload.Result.ResultType
+	fields[auditMetadataSource] = source
+
+	marshalled, err := marshal(payload)
+	if err != nil {
+		return err
+	}
+	fields[auditMetadataFullPayload] = marshalled
+
+	// log message can be anything here, since the important audit data are
+	// captured in the different log fields above.
+	entry.WithFields(fields).Info("see auditFullPayload field for full log data")
+	return nil
+}
+
+// Log captures information about the operation to be logged.
+type Log struct {
+	AzureEnvironment string
+	CallerIdentities []*schema.CallerIdentity
+	Category         schema.Category
+	CorrelationID    string
+	OperationID      string
+	OperationName    string
+	OperationResult  schema.Result
+	Region           string
+	Role             string
+	RoleInstance     string
+	TargetResources  []*schema.TargetResource
+}
+
+func (l *Log) toPayload() *schema.AuditPayload {
+	utcnow := time.Now().UTC()
+	payload := &schema.AuditPayload{
+		// Part-A. See schema/doc.go
+		EnvOS:                runtime.GOOS,
+		EnvVer:               ifxAuditVersion,
+		EnvName:              ifxAuditName,
+		EnvTime:              utcnow.Format(time.RFC3339),
+		EnvEpoch:             epoch,
+		EnvSeqNum:            nextSeqNum(),
+		EnvPopSample:         ifxAuditPopSample,
+		EnvFlags:             ifxAuditFlags,
+		EnvCV:                l.CorrelationID,
```

Comment:
just a question - what does CV stand for?

Reply:
Correlation vector - I don't know why it's called that either. I only know it's a string type that other projects has used to capture the correlation ID.

Reply:
Also, I don't know how much freedom I have to alter the name of the fields. There is a PR that I referred (see feature description), which has been reviewed and approved. There are some details (such as field names) that I just copy-and-pasted.